cagnotte_base/models/cagnotte.py

```diff
@@ -0,0 +1,202 @@
+        ondelete='restrict')
+    solde_cagnotte = fields.Float(
+        compute='_compute_solde_cagnotte',
+        store=True)
+    active = fields.Boolean(
+        default=True)
+    account_move_line_ids = fields.One2many(
+        comodel_name="account.move.line",
+        inverse_name="account_cagnotte_id",
+        string="Journal Items")
+    create_date = fields.Date(
+        default=fields.Date.today)
+
+    @api.multi
+    def _get_name(self):
+        self.ensure_one()
+        return '%s - %s' % (self.cagnotte_type_id.name, self.name)
+
+    @api.multi
+    def name_get(self):
+        """Add the type of cagnotte in the name"""
+        res = []
+        for cagnotte in self:
+            name = cagnotte._get_name()
+            res.append((cagnotte.id, name))
+        return res
+
+    @api.multi
+    @api.depends('account_move_line_ids.debit',
+                 'account_move_line_ids.credit')
+    def _compute_solde_cagnotte(self):
+        for cagnotte in self:
+            solde_cagnotte = 0
+            for move_line in cagnotte.account_move_line_ids:
+                solde_cagnotte += move_line.credit - move_line.debit
+            cagnotte.solde_cagnotte = solde_cagnotte
+
+    @api.multi
+    def create(self, vals):
+        if 'name' not in vals:
+            cagnotte_type = self.env['cagnotte.type'].browse(
+                vals['cagnotte_type_id'])
+            vals['name'] = cagnotte_type.sequence_id.next_by_id()
+        return super(AccountCagnotte, self).create(vals)
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    account_cagnotte_id = fields.Many2one(
+        comodel_name='account.cagnotte',
+        string='Cagnotte',
+        ondelete='restrict')
+
+    @api.model
+    def cagnotte_value(self, values):
+        """ If used account is on a cagnotte type,
+            create a cagnotte
+        """
+        if not values.get('account_cagnotte_id'):
+            if values.get('account_id') and values.get('credit', 0) > 0:
+                cagnotte_type = self.env['cagnotte.type'].search(
+                    [('account_id', '=', values['account_id'])])
+                if cagnotte_type:
+                    # create cagnotte
+                    values['account_cagnotte_id'] = \
+                        self.env['account.cagnotte'].create(
+                            {'cagnotte_type_id': cagnotte_type.id}).id
+        return values
+
+    @api.multi
+    def create(self, values):
+        vals = self.cagnotte_value(values)
+        return super(AccountMoveLine, self).create(vals)
+
+    @api.multi
+    def _check_cagnotte_account(self):
+        """ Account must correspond to cagnotte account
+        """
+        for line in self:
+            if line.account_cagnotte_id:
+                if line.account_cagnotte_id.cagnotte_type_id.account_id.id \
+                        != line.account_id.id:
+                    return False
+        return True
+
+    _constraints = [
```

Comment:
@Cedric-Pigeon Should be replaced by `@api.constrains` and a `validationError` in the method.